homeassistant/components/plex/__init__.py

```diff
@@ -1 +1,195 @@
-"""The plex component."""
+"""Support to embed Plex."""
+import logging
+
+import plexapi.exceptions
+import voluptuous as vol
+
+from homeassistant.components.discovery import SERVICE_PLEX
+from homeassistant.components.media_player import DOMAIN as MP_DOMAIN
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_TOKEN,
+    CONF_URL,
+    CONF_VERIFY_SSL,
+)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import discovery
+from homeassistant.util.json import load_json, save_json
+
+from .const import (
+    CONF_USE_EPISODE_ART,
+    CONF_SHOW_ALL_CONTROLS,
+    CONF_REMOVE_UNAVAILABLE_CLIENTS,
+    CONF_CLIENT_REMOVE_INTERVAL,
+    DEFAULT_HOST,
+    DEFAULT_PORT,
+    DEFAULT_SSL,
+    DEFAULT_VERIFY_SSL,
+    DOMAIN as PLEX_DOMAIN,
+    PLATFORMS,
+    PLEX_CONFIG_FILE,
+    PLEX_MEDIA_PLAYER_OPTIONS,
+    SERVERS,
+)
+from .server import PlexServer
+
+MEDIA_PLAYER_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_USE_EPISODE_ART, default=False): cv.boolean,
+        vol.Optional(CONF_SHOW_ALL_CONTROLS, default=False): cv.boolean,
+        vol.Optional(CONF_REMOVE_UNAVAILABLE_CLIENTS, default=True): cv.boolean,
+        vol.Optional(CONF_CLIENT_REMOVE_INTERVAL, default=600): cv.positive_int,
+    }
+)
+
+SERVER_CONFIG_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
+        vol.Optional(CONF_TOKEN): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+        vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+        vol.Optional(MP_DOMAIN, default={}): MEDIA_PLAYER_SCHEMA,
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema({PLEX_DOMAIN: SERVER_CONFIG_SCHEMA}, extra=vol.ALLOW_EXTRA)
+
+_CONFIGURING = {}
+_LOGGER = logging.getLogger(__package__)
+
+
+def setup(hass, config):
+    """Set up the Plex component."""
+
+    def server_discovered(service, info):
+        """Pass back discovered Plex server details."""
+        _LOGGER.debug("Discovered Plex server: %s:%s", info["host"], info["port"])
+        setup_plex(discovery_info=info)
+
+    def setup_plex(config=None, discovery_info=None, configurator_info=None):
```

Comment:
We'll be able to unnest all these functions when we implement config entries.

Reply:
I can't wait 😊. Anything to address in this PR?

Reply:
No. :+1: